### DISCUSSION TITLE
运动控制部分重新实现

### DISCUSSION THREAD

**Discussion by zbwu** [MiRoboticsLab/cyberdog_ros2#31] [Ideas]
有同学对这个感兴趣的吗？

因为官方开源计划的变化，这部分将不会开源，重新实现看来有必要性了。

或者可以先来一个可行性探索。

**Comment by homalozoa** [comment]
有

**Comment by c-tx** [comment]
+1先搞一搞运动学动力学

**Comment by liseri** [comment]
有；提供另一个思路，运动黑盒也可以玩，摸清通信协议就可以；我想先把这部分实现；

**Reply by a52521255** [reply]
摸清四路can么

**Comment by zbwu** [comment]
找到了一些cyberdog的物理参数

### Property ###
body_length: 4.70719993114471435546875000000000e-01
body_width: 1.00000001490116119384765625000000e-01
body_height: 1.00000001490116119384765625000000e-01
body_mass: 6.51999998092651367187500000000000e+00
abad_gear_ratio: 6.00000000000000000000000000000000e+00
hip_gear_ratio: 6.00000000000000000000000000000000e+00
knee_gear_ratio: 9.00000000000000000000000000000000e+00
abad_link_length: 1.07150003314018249511718750000000e-01
hip_link_length: 2.00000002980232238769531250000000e-01
knee_link_length: 2.16999992728233337402343750000000e-01
knee_link_y_offset: 0.00000000000000000000000000000000e+00
max_leg_length: 4.09000009298324584960937500000000e-01
motor_kt: 5.00000007450580596923828125000000e-02
motor_r: 1.72999992966651916503906250000000e-01
battery_v: 2.40000000000000000000000000000000e+01
motor_tau_max: 3.00000000000000000000000000000000e+00
joint_damping: 9.99999977648258209228515625000000e-03
joint_dry_friction: 2.00000002980232238769531250000000e-01

### Spatial Inertia ###
abad: [ 5.090000e-01, -2.290500e-03, 9.162000e-04, -2.545000e-03, 3.809741e-04, 6.938322e-04, 4.733564e-04, 5.071000e-06, -1.165250e-05, 1.252290e-05 ]
hip: [ 6.640000e-01, -1.925600e-03, -2.217760e-02, -1.235040e-02, 3.337649e-03, 2.638502e-03, 1.309316e-03, -9.303330e-06, -1.928162e-04, -7.150375e-07 ]
knee: [ 1.140000e-01, 8.094000e-04, 4.560000e-06, -1.281474e-02, 1.455305e-03, 2.152152e-03, 7.054469e-04, 5.125896e-07, 8.388462e-05, -3.237600e-08 ]
abad_rotor: [ 5.500000e-02, 0.000000e+00, 0.000000e+00, 0.000000e+00, 6.300000e-05, 3.300000e-05, 3.300000e-05, 0.000000e+00, 1.311342e-12, 0.000000e+00 ]
hip_rotor: [ 5.500000e-02, 0.000000e+00, 0.000000e+00, 0.000000e+00, 3.300000e-05, 6.300000e-05, 3.300000e-05, -1.311342e-12, 0.000000e+00, 0.000000e+00 ]
knee_rotor: [ 5.500000e-02, 0.000000e+00, 0.000000e+00, 0.000000e+00, 3.300000e-05, 6.300000e-05, 3.300000e-05, -1.311342e-12, 0.000000e+00, 0.000000e+00 ]
body: [ 6.520000e+00, 1.284440e-01, -1.564800e-03, -5.737600e-03, 3.205143e-02, 1.370674e-01, 1.494587e-01, 5.662298e-05, 2.728031e-03, -2.321734e-04 ]

### Location ###
abad: [ 2.353600e-01, 5.000000e-02, 0.000000e+00 ]
abad_rotor: [ 2.353600e-01, 5.000000e-02, 0.000000e+00 ]
hip: [ 0.000000e+00, 1.071500e-01, 0.000000e+00 ]
hip_rotor: [ 0.000000e+00, 4.000000e-02, 0.000000e+00 ]
knee: [ 0.000000e+00, 0.000000e+00, -2.000000e-01 ]
knee_rotor: [ 0.000000e+00, 0.000000e+00, 0.000000e+00 ]

**Comment by zbwu** [comment]
cyberdog的物理参数，更新一个版本，更便于粘贴到代码里面，数据精度改为double。

/* Parameter */
cheetah._bodyLength = 4.707199931144714e-01
cheetah._bodyWidth = 1.000000014901161e-01
cheetah._bodyHeight = 1.000000014901161e-01
cheetah._bodyMass = 6.519999980926514e+00
cheetah._abadGearRatio = 6.000000000000000e+00
cheetah._hipGearRatio = 6.000000000000000e+00
cheetah._kneeGearRatio = 9.000000000000000e+00
cheetah._abadLinkLength = 1.071500033140182e-01
cheetah._hipLinkLength = 2.000000029802322e-01
cheetah._kneeLinkLength = 2.169999927282333e-01
cheetah._kneeLinkY_offset = 0.000000000000000e+00
cheetah._maxLegLength = 4.090000092983246e-01
cheetah._motorKT = 5.000000074505806e-02
cheetah._motorR = 1.729999929666519e-01
cheetah._batteryV = 2.400000000000000e+01
cheetah._motorTauMax = 3.000000000000000e+00
cheetah._jointDamping = 9.999999776482582e-03
cheetah._jointDryFriction = 2.000000029802322e-01

/* Spatial Inertia */
MassProperties<T> abadMassProperties << 5.090000033378601e-01, -2.290499862283468e-03, 9.161999914795160e-04, -2.545000053942204e-03, 3.809741465374827e-04, 6.938322330825031e-04, 4.733563982881606e-04, 5.070999577583279e-06, -1.165249886980746e-05, 1.252289894182468e-05;
cheetah._abadInertia = SpatialInertia<T>(abadMassProperties);
MassProperties<T> hipMassProperties << 6.639999747276306e-01, -1.925599877722561e-03, -2.217759937047958e-02, -1.235039997845888e-02, 3.337649162858725e-03, 2.638501580804586e-03, 1.309316023252904e-03, -9.303330443799496e-06, -1.928161655087024e-04, -7.150374585762620e-07;
cheetah._hipInertia = SpatialInertia<T>(hipMassProperties);
MassProperties<T> kneeMassProperties << 1.140000000596046e-01, 8.093999931588769e-04, 4.559999979392160e-06, -1.281473971903324e-02, 1.455305144190788e-03, 2.152151660993695e-03, 7.054469315335155e-04, 5.125896223034943e-07, 8.388462447328493e-05, -3.237600054717404e-08;
cheetah._kneeInertia = SpatialInertia<T>(kneeMassProperties);
MassProperties<T> abadRotorMassProperties << 5.499999970197678e-02, 0.000000000000000e+00, 0.000000000000000e+00, 0.000000000000000e+00, 6.299999949987978e-05, 3.300000025774352e-05, 3.300000025774352e-05, 0.000000000000000e+00, 1.311341551839262e-12, 0.000000000000000e+00;
cheetah._abadRotorInertia = SpatialInertia<T>(abadRotorMassProperties);
MassProperties<T> hipRotorMassProperties << 5.499999970197678e-02, 0.000000000000000e+00, 0.000000000000000e+00, 0.000000000000000e+00, 3.300000025774352e-05, 6.299999949987978e-05, 3.300000025774352e-05, -1.311341551839262e-12, 0.000000000000000e+00, 0.000000000000000e+00;
cheetah._hipRotorInertia = SpatialInertia<T>(hipRotorMassProperties);
MassProperties<T> kneeRotorMassProperties << 5.499999970197678e-02, 0.000000000000000e+00, 0.000000000000000e+00, 0.000000000000000e+00, 3.300000025774352e-05, 6.299999949987978e-05, 3.300000025774352e-05, -1.311341551839262e-12, 0.000000000000000e+00, 0.000000000000000e+00;
cheetah._kneeRotorInertia = SpatialInertia<T>(kneeRotorMassProperties);
MassProperties<T> bodyMassProperties << 6.519999980926514e+00, 1.284440010786057e-01, -1.564799924381077e-03, -5.737599916756153e-03, 3.205142542719841e-02, 1.370674073696136e-01, 1.494587212800980e-01, 5.662297917297110e-05, 2.728030551224947e-03, -2.321734355064109e-04;
cheetah._bodyInertia = SpatialInertia<T>(bodyMassProperties);

/* Location */
cheetah._abadLocation << 2.353599965572357e-01, 5.000000074505806e-02, 0.000000000000000e+00;
cheetah._abadRotorLocation << 2.353599965572357e-01, 5.000000074505806e-02, 0.000000000000000e+00;
cheetah._hipLocation << 0.000000000000000e+00, 1.071500033140182e-01, 0.000000000000000e+00;
cheetah._hipRotorLocation << 0.000000000000000e+00, 3.999999910593033e-02, 0.000000000000000e+00;
cheetah._kneeLocation << 0.000000000000000e+00, 0.000000000000000e+00, -2.000000029802322e-01;
cheetah._kneeRotorLocation << 0.000000000000000e+00, 0.000000000000000e+00, 0.000000000000000e+00;

**Comment by zbwu** [comment]
imu和power mcu的协议也补完了

和上面的物理参数一同到下面repo

https://github.com/zbwu/cyberdog_misc.git

**Reply by liseri** [reply]
为啥不是从lcm读呢

**Comment by fan-ziqi** [comment]
小米铁蛋仿真（基于宇树环境）https://github.com/fan-ziqi/cyberdog_sim
由于宇树控制代码用的是单刚体简化模型，简化模型的质量、转动惯量、重心位置均无法获得，只给了经验值。若需要考虑所有构件的质量属性则需要移植WBC。
